server/src/frameworks/Foundry/FoundryProject.ts

```diff
@@ -0,0 +1,286 @@
+import Project from "../base/Project";
+import { Remapping } from "./Remapping";
+
+interface SolFileDetails {
+  path: string;
+  pragmas: string[];
+}
+
+export default class FoundryProject extends Project {
+  public priority = 1;
+  public sourcesPath!: string;
+  public testsPath!: string;
+  public remappings: Remapping[] = [];
+  public initializeError?: string;
+  public configSolcVersion?: string;
+
+  constructor(
+    serverState: ServerState,
+    basePath: string,
+    public configPath: string
+  ) {
+    super(serverState, basePath);
+  }
+
+  public id(): string {
+    return this.configPath;
+  }
+
+  public frameworkName(): string {
+    return "Foundry";
+  }
+
+  public async initialize(): Promise<void> {
+    try {
+      const config = JSON.parse(
+        await runCmd("~/.foundry/bin/forge config --json", this.basePath)
+      );
+      this.sourcesPath = path.join(this.basePath, config.src);
+      this.testsPath = path.join(this.basePath, config.test);
+      this.configSolcVersion = config.solc || undefined; // may come as null otherwise
+
+      const rawRemappings = await runCmd(
+        "~/.foundry/bin/forge remappings",
+        this.basePath
+      );
+      this.remappings = this._parseRemappings(rawRemappings);
+    } catch (error: any) {
+      this.serverState.logger.error(error.toString());
+
+      switch (error.code) {
+        case 127:
+          this.initializeError =
+            "Couldn't run `forge`. Please check that your foundry installation is correct.";
+          break;
+        case 134:
+          this.initializeError =
+            "Running `forge` failed. Please check that your foundry.toml file is correct.";
+          break;
+        default:
+          this.initializeError = `Unexpected error while running \`forge\`: ${error}`;
+      }
+    }
+
+    return;
+  }
+
+  public async fileBelongs(uri: string) {
+    if (this.initializeError === undefined) {
+      // Project was initialized correctly, then check contract is inside source or test folders
+      return [this.sourcesPath, this.testsPath].some((dir) =>
+        directoryContains(dir, uri)
+      );
+    } else {
+      // Project could not be initialized. Claim all files under base path to avoid them being incorrectly assigned to other projects
+      return directoryContains(this.basePath, uri);
+    }
+  }
+
+  public resolveImportPath(file: string, importPath: string) {
+    try {
+      let transformedPath = importPath;
+
+      if (!importPath.startsWith(".")) {
+        for (const { from, to } of this.remappings) {
+          if (importPath.startsWith(from)) {
+            transformedPath = path.join(to, importPath.slice(from.length));
+          }
+        }
+      }
+      const resolvedPath = require.resolve(transformedPath, {
+        paths: [fs.realpathSync(path.dirname(file))],
+      });
+
+      return toUnixStyle(fs.realpathSync(resolvedPath));
+    } catch (error) {
```

Comment:
The `resolveImportPath` function is designed to return either a resolved path or undefined. In the implementation, we perform a try/catch because `require.resolve` throws an error when it fails to resolve. I don't think it makes sense to report these kind of errors because it happens on each invalid import. The current analysis strategy involves just ignoring the imports that can't be crawled because they can't be resolved. From the validation standpoint, all invalid imports are shown as diagnostics.

Reply:
If a file imports a path `/some/path.sol`, and the resolveImportPath succeeds, but the analysis of that file `/some/path.sol` fails in the recursive analyser, does the starting file also fail analysis?

Reply:
No, in that case everything work as expected, just that the invalid import (from /some/path.sol) is not crawled. But from the initial file to /some/path.sol everything is tied correctly

Reply:
Hmmm, I prefer that we do nothing rather than doing the wrong thing.

If we do a rename, but analysis from the file failed on an import, the rename could be wrong. Better to just fail the rename because the analysis failed.

What do you think?

Reply:
That is a fair point. Lets not add "fix everything" to this PR (I'll add "fix everything" to the backlog).